elks/arch/i86/drivers/char/serfast.S

```diff
@@ -27,7 +25,11 @@ _irq_com1:
 	push	%dx
 	push	%ds
 
-	mov	%cs:ds_kernel,%ds	// recover kernel data segment
+	// Recover kernel data segment
+	// Was pushed by the CALLF of the dynamic handler
+	mov	%sp,%bx
+	mov	%ss:12(%bx),%ds
+
```

Comment:
Thanks for the update here. So, the fast interrupt routines will still use a dynamic handler trampoline? For some reason I was thinking they were going to be routed directly from low-core vector. 

But old code still went through _irqtable stub, and we will likely always require a way to get kernel DS, and this removes ds_kernel kluge.

Did you by chance test this to make sure the 12 offset is correct?

To make sure I fully understand: since SS can be absolutely anything at the time of fast serial interrupt, we are using SS here to access the original saved far call segment pushed by hardware interrupt to get to the trampoline? And that trampoline segment will always be the kernel DS?

If so, that means your comment "Was pushed by the CALLF of the dynamic handler" isn't correct: the data segment was pushed by the hardware interrupt prior to execution of the CALLF in the dynamic handler...

It seems there are some nice benefits to using the trampoline in kernel data after all. Since old method used trampoline in code segment, it was not easy to get at kernel data segment without kluge. This way, we get kernel data easily, but have no idea where kernel code segment is, and don't care. Correct?

Reply:
Yes, the fast static handlers are also called by the 'trampoline', because now the kernel DS is pushed by that trampoline and not stored anymore in `ds_kernel`. Before the changes, there were a JMP FAR and a CS prefixed MOV from memory. 

Now there are a CALL FAR (pushing two words), a MOV from register to register (very fast as CPU internal), and a SS prefixed MOV from memory (memory offset has no impact). Therefore the path is a bit slower because of two more pushes, but I think it won't change the game.

In slow or fast interrupt, whatever the initial SS:SP, the HW first pushes 6 bytes (FL, CS, IP) and JMP FAR to the trampoline at kernel DS:xxx (in the heap). Then the trampoline CALL FAR actually pushes 4 bytes (offset of `irq`, DS) and JMP FAR to the static handler. On static handler routine entry, the stack is:
```
pushed by the tampoline:
    irq offset  <-- SS:SP
    kernel DS
pushed by the HW:
    next IP after interrupt
    interrupted CS
    interrupted FL
```

That makes both the kernel DS and the pointer to IRQ number available on the stack for the static handler. Also the comment is correct, kernel DS was pushed by the trampoline.

Reply:
Thanks for the comments. I get it now, and comment is correct. 

New method, found by having to add CALLF in trampoline, turns out to have nice benefits for ELKS interrupt handling. I agree now that overall solution is superior, since stack frame contains all information needed for interrupt handler, whereas before it did not and required cs:ds_kernel kluge. Nice work!